pydantic/_internal/_namespace_utils.py

```diff
@@ -0,0 +1,237 @@
+            with `TypeAdapter`, where the "type" being analyzed is a simple annotation.
+        fallback_namespace: A namespace to use as a last resort if a name wasn't found
+            in the locals nor the globals. This is mainly used when rebuilding a core
+            schema for a Pydantic model or dataclass, where the parent namespace is used.
+        override_namespace: A namespace to use first when resolving forward annotations.
+            This is mainly used when rebuilding a core schema for a Pydantic model or
+            dataclass, where an explicit namespace can be provided to resolve annotations
+            that failed to resolve during the initial schema building.
+
+    Example:
+        ```python lint="skip" test="skip"
+        ns_resolver = NsResolver(
+            namespaces_tuple=NamespacesTuple({'my_global': 1}, {}),
+            fallback_namespace={'fallback': 2, 'my_global': 3},
+        )
+
+        ns_resolver.types_namespace
+        #> NamespacesTuple({'my_global': 1}, {})
+
+
+        class MyType:
+            some_local = 4
+
+
+        with ns_resolver.push(MyType):
+            ns_resolver.types_namespace
+            #> NamespacesTuple({'my_global': 1, 'fallback': 2}, {'some_local': 4})
+        ```
+    """
+
+    def __init__(
+        self,
+        namespaces_tuple: NamespacesTuple | None = None,
+        fallback_namespace: MappingNamespace | None = None,
+        override_namespace: MappingNamespace | None = None,
+    ) -> None:
+        self._base_ns_tuple = namespaces_tuple or NamespacesTuple({}, {})
+        self._fallback_ns = fallback_namespace
+        self._override_ns = override_namespace
+        self._types_stack: list[type[Any]] = []
+
+    @cached_property
+    def types_namespace(self) -> NamespacesTuple:
+        """The current global and local namespaces to be used for annotations evaluation."""
+        if not self._types_stack:
+            # TODO do we want to merge fallback and override ns here?
+            return self._base_ns_tuple
+
+        typ = self._types_stack[-1]
+
+        globalns = get_module_ns_of(typ)
+        # TODO check len(self._types_stack) == 1? As we specified,
+        # fallback_namespace should only be used for the class being rebuilt.
+        # However, it might be good to keep this for now to avoid having too
+        # many breaking changes.
+        if self._fallback_ns is not None:
+            globalns = {**self._fallback_ns, **globalns}
+
+        locals_list: list[MappingNamespace] = [vars(typ), {typ.__name__: typ}]
+        if self._override_ns is not None:
+            locals_list.append(self._override_ns)
+
+        return NamespacesTuple(globalns, LazyLocalNamespace(*locals_list))
+
+    @contextmanager
+    def push(self, typ: type[Any], /) -> Generator[None]:
+        """Push a type to the stack."""
+        self._types_stack.append(typ)
+        # Reset the cached property:
+        self.__dict__.pop('types_namespace', None)
```

Comment:
You could have something like:

```python
from dataclasses import dataclass

@dataclass
class A:
    a: 'F1'
    b: 'F2'

class Model(BaseModel):
    a: A

# Model is still incomplete because of unresolved annotations

F1 = int
f2 = str

Model.model_rebuild()
# re-evaluating `A.a` will access the `types_namespace` property, compute it,
# re-evaluating `A.b` will access the `types_namespace` property, using the cached value
```

Reply:
It doesn't add that much performance benefit, but still good to have considering the cache invalidation logic is pretty simple

Reply:
Makes sense, thanks for the example. Marking as resolved 👍